rllib/algorithms/dt/segmentation_buffer.py

```diff
@@ -0,0 +1,215 @@
+        # same episode is concatenated together twice (which is okay).
+        if episode.env_steps() > self.max_ep_len:
+            logger.warning(
+                f"The maximum rollout length is {self.max_ep_len} but we tried to add a"
+                f"rollout of {episode.env_steps()} steps to the SegmentationBuffer. "
+                f"This could be due to incorrect data in the dataset (need fixing), "
+                f"or random shuffling that caused a duplicate rollout (no need to fix)."
+            )
+            episode = episode[: self.max_ep_len]
+
+        if len(self._buffer) < self.capacity:
+            self._buffer.append(episode)
+        else:
+            # TODO: replace proportional to episode length
+            replace_ind = np.random.randint(0, self.capacity)
+            self._buffer[replace_ind] = episode
+
+    def sample(self, batch_size: int) -> SampleBatch:
+        """Sample segments from the buffer.
+
+        Args:
+            batch_size: number of segments to sample.
+
+        Returns:
+            SampleBatch of segments with keys and shape {
+                OBS: [batch_size, max_seq_len, obs_dim],
+                ACTIONS: [batch_size, max_seq_len, act_dim],
+                RETURNS_TO_GO: [batch_size, max_seq_len + 1, 1],
+                T: [batch_size, max_seq_len],
+                ATTENTION_MASKS: [batch_size, max_seq_len],
+            }
+        """
+        samples = [self._sample_single() for _ in range(batch_size)]
+        return concat_samples(samples)
+
+    def _sample_single(self) -> SampleBatch:
+        # TODO: sample proportional to episode length
+        # Sample a random episode from the buffer and then sample a random
+        # segment from that episode.
+        buffer_ind = np.random.randint(0, len(self._buffer))
+        episode = self._buffer[buffer_ind]
+        ep_len = episode[SampleBatch.OBS].shape[0]
+        # This offset accounts for either the normal case when max_seq_len is shorter
+        # than the episode length, but also occasionally an episode will be shorter
+        # than the context length (max_seq_len).
+        offset = min(self.max_seq_len, ep_len)
+        # We allow si to be negative (for now) because we want segments that only
+        # contains the first few transitions (and padd the rest),
+        # for example [0, 0, 0, 0, 0, 0, R0, s0, a0].
+        si = np.random.randint(-offset + 1, ep_len - offset + 1)
+        ei = si + offset
+        # but for actual segmenting we don't want starting index to be negative.
+        si = max(si, 0)
+
+        obs = episode[SampleBatch.OBS][si:ei]
+        actions = episode[SampleBatch.ACTIONS][si:ei]
+        # Note that returns-to-go needs an extra elem as the target for the last action.
+        returns_to_go = episode[SampleBatch.RETURNS_TO_GO][si : ei + 1].reshape(-1, 1)
+
+        length = obs.shape[0]
+        timesteps = np.arange(si, si + length, dtype=np.int32)
```

Comment:
**TODO** for timesteps can we just use `SampleBatch.T`, the algorithm should postprocess the returned sample batch from the dataset reader to compute rtg as well as append the dataset with timesteps key if it doesn' t already exist.

Reply:
It seems unnecessary to store arrays that we can just generate, but I think the main confusion might be that RLlib uses -1 as its first index I think, so even if SampleBatch.T exists we might have to overwrite it in postprocess_trajectory. Correct me if I'm wrong though.